### ISSUE TITLE
[question] Local development without real DNS resolution with local keycloak and oauth2-proxy

### ISSUE THREAD

**Issue by roderik** [oauth2-proxy/oauth2-proxy#1739]
## Expected Behavior

We are developing locally in a KinD cluster we can spin up and down in seconds. It has an nginx-ingress controller and we combine it with hostnames like *.fuf.me (our xip.io etc) that resolve to localhost by default. This works great for accessing ingress exposed services.

In our stack we deploy both keycloak and oauth2-proxy. Keycloak is on https://auth.app.fuf.me, a service might be on https://app.fuf.me/grafana and oauth2-proxy is on https://app.fuf.me/oauth2/callback

Grafana's annotations are

```yaml
    ingress:
      annotations:
        kubernetes.io/tls-acme: 'true'
        cert-manager.io/cluster-issuer: selfsigned-cluster-issuer
        nginx.ingress.kubernetes.io/auth-response-headers: "Authorization, x-auth-request-user, x-auth-request-email"
        nginx.ingress.kubernetes.io/auth-signin: https://$host/oauth2/start?rd=$escaped_request_uri
        nginx.ingress.kubernetes.io/auth-url: https://$host/oauth2/auth
        nginx.ingress.kubernetes.io/configuration-snippet: |
          auth_request_set $name_upstream_1 $upstream_cookie__oauth2_proxy_1;
          
          access_by_lua_block {
            if ngx.var.name_upstream_1 ~= "" then
              ngx.header["Set-Cookie"] = "_oauth2_proxy_1=" .. ngx.var.name_upstream_1 .. ngx.var.auth_cookie:match("(; .*)")
            end
          }
      hosts:
        - 'app.fuf.me'
      path: '/grafana'
      tls:
        - secretName: app.fuf.me-tls
          hosts:
            - app.fuf.me
```

There is a bit of a challenge using these domain names however inside of the cluster. Specifically with oidc discovery, as everywhere, also inside of the container, the urls resolve to localhost.

I've gotten quite far by disabling discovery and selectively using the internal service name, apart from the login url

```yaml
oauth2-proxy:
  extraArgs:
    redirect-url: 'https://app.fuf.me/oauth2/callback'
    oidc-issuer-url: 'http://auth-keycloak/realms/x'
    reverse-proxy: "true"
    insecure-oidc-skip-issuer-verification: "true"
    skip-oidc-discovery: "true"
    login-url: "https://auth.app.fuf.me/realms/x/protocol/openid-connect/auth"
    redeem-url: "https://auth-keycloak/realms/x/protocol/openid-connect/token"
    oidc-jwks-url: "https://auth-keycloak/realms/x/protocol/openid-connect/certs"
    validate-url: "https://auth-keycloak/auth/realms/x/protocol/openid-connect/userinfo"
    silence-ping-logging: "true"
    show-debug-on-error: "true"
  ingress:
    annotations:
      kubernetes.io/tls-acme: 'true'
      cert-manager.io/cluster-issuer: selfsigned-cluster-issuer
    path: /oauth2
    hosts:
      - 'app.fuf.me'
    tls:
      - secretName: app.fuf.me-tls
        hosts:
          - app.fuf.me
```          

At this point I can visit /grafana, get redirected to Keycloak, sign in and get redirected back to the callback/redirect url in my browser, but there it stalls and times out.

Example redirect url

```
https://app.fuf.me/oauth2/callback?state=ikixkWNrtowPaNhDGHCrqXKz2odREi-jUoIhLGhiRWM%3A%2Fgrafana&session_state=6edd1cdb-23b1-4f91-bed5-977b0c4c741c&code=23669ac6-56db-4b90-8278-5ad1641644c3.6edd1cdb-23b1-4f91-bed5-977b0c4c741c.f5264178-53b4-4626-a075-a4b2c8bfe679
```

the oauth2-proxy logs show me 

```
[2022/07/27 07:40:28] [oauthproxy.go:768] Error redeeming code during OAuth2 callback: token exchange failed: context canceled                                                                                                                                                                             
172.18.0.1 - abea16f003d04fc222f165c02dcf5d92 - - [2022/07/27 07:39:28] app.fuf.me GET - "/oauth2/callback?state=ikixkWNrtowPaNhDGHCrqXKz2odREi-jUoIhLGhiRWM%3A%2Fgrafana&session_state=6edd1cdb-23b1-4f91-bed5-977b0c4c741c&code=23669ac6-56db-4b90-8278-5ad1641644c3.6edd1cdb-23b1-4f91-bed5-977b0c4c74
1c.f5264178-53b4-4626-a075-a4b2c8bfe679" HTTP/1.1 "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.0 Safari/605.1.15" 500 2793 59.999 
```

And the ingress controller shows a timeout

```
2022/07/27 07:39:28 [error] 643#643: *484464 upstream timed out (110: Operation timed out) while reading response header from upstream, client: 172.18.0.1, server: bpaas.fuf.me, request: "GET /oauth2/callback?state=ikixkWNrtowPaNhDGHCrqXKz2odREi-jUoIhLGhiRWM%3A%2Fgrafana&session_state=6edd1cdb-23b1
-4f91-bed5-977b0c4c741c&code=23669ac6-56db-4b90-8278-5ad1641644c3.6edd1cdb-23b1-4f91-bed5-977b0c4c741c.f5264178-53b4-4626-a075-a4b2c8bfe679 HTTP/2.0", upstream: "http://10.244.0.31:4180/oauth2/callback?state=ikixkWNrtowPaNhDGHCrqXKz2odREi-jUoIhLGhiRWM%3A%2Fgrafana&session_state=6edd1cdb-23b1-4f91-b
ed5-977b0c4c741c&code=23669ac6-56db-4b90-8278-5ad1641644c3.6edd1cdb-23b1-4f91-bed5-977b0c4c741c.f5264178-53b4-4626-a075-a4b2c8bfe679", host: "app.fuf.me"  
```

Keycloak (bitnami version) is completely silent.

Which brings me to the issue where apparently the redeem-url is not working as expected but nothing is showing me an actionable error message. 

So long story short, anyone have a clue on how to get this working or find out why the token exchange fails? 

I looked at alternative dns solutions like in Minikube, but all options are a huge pain to get configure and install if the development flow recreates the cluster 20 times a day.

**Comment by JoelSpeed** [comment]
Have you tried execing into the OAuth2 Proxy pod and using curl to check the redeem URL? If token exchange is failing then it can't request the redeem URL. 

Are the applications deployed in the same namespace? If not you should use the fully qualified service name
